### PR TITLE
Parallelize flutter platform compilation by number of cores available.

### PR DESCRIPTION
**Pull request by aam**
Gives ~3x performance boost on "standard" 28 core linux desktop

before
```
$FH/flutter/packages/flutter$ $FH/flutter/bin/flutter test
02:52 +3093 ~25: All tests passed!                                                                                                                                                                                
$FH/flutter/bin/flutter test  930.92s user 130.69s system 606% cpu 2:55.15 total
```

after
```
$FH/flutter/packages/flutter$ $FH/flutter/bin/flutter test
00:47 +3093 ~25: All tests passed!                                                                                                                                                                                
$FH/flutter/bin/flutter test  1144.62s user 127.94s system 2553% cpu 49.843 total
```

"Inspired" by @jensjoha 's https://github.com/flutter/flutter/pull/15511

### CODE REVIEW

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 1]
```diff
@@ -80,6 +80,7 @@ Future<int> runTests(
     trackWidgetCreation: trackWidgetCreation,
     updateGoldens: updateGoldens,
     projectRootDirectory: fs.currentDirectory.uri,
+    concurrency: concurrency
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/test/runner.dart]: nit: trailing comma

File: packages/flutter_tools/lib/src/test/runner.dart [hunk 2]
```diff
@@ -80,6 +80,7 @@ Future<int> runTests(
...
   );
 
   // Make the global packages path absolute.
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/test/runner.dart]: tests?

**Comment by Hixie** [comment]
How does this interact with the `test` package's ideas of concurrency? Will we be running N tests while compiling N tests? or do we only ever compile a test when we are about to start a test?

**Comment by aam** [comment]
> How does this interact with the test package's ideas of concurrency? Will we be running N tests while compiling N tests? or do we only ever compile a test when we are about to start a test?

Every test needs to be compiled first, so there is a compilation step that precedes test execution. But those two phases are only loosely connected with potentially equal number of tests being run as compilers compiling (next batch of) tests.
The way it set up right now(I'm reusing testing `concurrency` option) it will have equal number of compilers created and running as number of concurrently running tests. With default `--concurrency` value of `numberOfProcessors - 2` that seem to have worked fine(3x improvement over current single compiler set up), but we can also update default for `--concurrency` to `numberOfProcessors/2`.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 3]
```diff
@@ -240,57 +239,68 @@ class _Compiler {
...
+      ResidentCompiler compiler;
+      String outputDill;
+      bool firstCompile = false;
+      if (idleCompilers.isEmpty && compilers < concurrency ) {
+        // Create compiler
+        compiler = createCompiler();
+        firstCompile = true;
+        compilers++;
+        outputDill = outputDillDirectory.childFile('output_$compilers.dill').path;
+        printTrace('Compiler will use the following file as its incremental dill file: $outputDill');
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: We are going to see multiples of these trace messages depending on the number of compilers being used, maybe the message should also say which compiler number it is referring to
'Compiler($compilers) will use ...'

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 4]
```diff
@@ -240,57 +239,68 @@ class _Compiler {
...
+            request.path);
+        final String outputPath = compilerOutput?.outputFilename;
+
+        // In case compiler didn't produce output or reported compilation
+        // errors, pass [null] upwards to the consumer and shutdown the
+        // compiler to avoid reusing compiler that might have gotten into
+        // a weird state.
+        if (outputPath == null || compilerOutput.errorCount > 0) {
+          request.result.complete(null);
+          await _shutdown(compiler);
```
Comment (a-siva) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: if we get into this path, aren't we supposed to exit the While loop in oder to go back and create a new instance of the compiler, or create a fresh instance of the compiler here?

**Review by a-siva — APPROVED** [review]
LGTM with one comment regarding the failure path.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 4 above)
Comment (aam) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Yes, good point!

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 5]
```diff
@@ -240,57 +239,69 @@ class _Compiler {
...
-            request.result.complete(kernelReadyToRun.path);
-            compiler.accept();
-            compiler.reset();
+      ResidentCompiler compiler;
+      String outputDill;
+      bool firstCompile = false;
+      if (idleCompilers.isEmpty && compilers < concurrency ) {
+        // Create compiler
+        compiler = createCompiler();
+        firstCompile = true;
```
Comment (jensjoha) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart]: I think this should only be done the very first time...

Comment (jensjoha) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Also, can funny stuff happen now, where the test cache is re-written while it's being loaded by another compiler?

**Review by jensjoha — COMMENTED** [review]
For my apparently non-standard 12-core (6 cores with hyper-threading I think) I get these results:

Before:

02:39 +3097 ~25: All tests passed!
02:32 +3097 ~25: All tests passed!
02:30 +3097 ~25: All tests passed!

With patch @ https://patch-diff.githubusercontent.com/raw/flutter/flutter/pull/21763.diff

01:57 +3097 ~25: All tests passed!
01:56 +3097 ~25: All tests passed!
01:56 +3097 ~25: All tests passed!

A few other tests:
With ~/ 4: 02:42 +3097 ~25: All tests passed! (i.e. slower than before!)
With ~/ 6: 03:40 +3097 ~25: All tests passed! (i.e. much slower than before!)
With ~/ 12: 07:16 +3097 ~25: All tests passed! (i.e. much much slower than before!)


Something else to consider:
As I recall when looking into the test package and how it does the concurrency thing it uses the provided concurrency as a limit of how many test starts running at once, but it does the waiting *after* the whole thing has been loaded into the VM (which I suppose made sense in the dart1 world where everything was compiled lazily, but I'm not quite sure in the dart2 world where the worlds is compiled (and loaded into the vm) up front). It "prepares" (or whatnot) more than this number (say twice that? (I really don't recall)), i.e. scedules more than the given concurrency, and potentially loads more VMs too (if the resident compiler(s) can keep up). We might thus both use half the cores compiling, all the cores loading VMs (but not yet starting the actual test) and then again half the cores to actually running the tests. Given that I'll try this: Set the concurrency as before (cores - 2) and seperatly limit the number of resident compilers:

cores - 2 and "if (idleCompilers.isEmpty && compilers < 2)": 01:54 +3097 ~25: All tests passed!
cores - 2 and "if (idleCompilers.isEmpty && compilers < 3)": 01:55 +3097 ~25: All tests passed!
cores - 3 and "if (idleCompilers.isEmpty && compilers < 3)": 01:54 +3097 ~25: All tests passed!

Also worth noting here is that - again, this is on my machine - with 2 compilers they basically both use 100% CPU (according to htop) all the time. With 3 compilers it goes down to ~50-80%.

So in "conclusion":
a) I'm not seing anywhere near the speedup you're seeing on my fewer-cores-cpu.
b) It becomes slightly (though probably not statisticly significantly) faster by doing this somewhat differently.
c) I'm not sure the added complexity is worth it; nor the added memory-overhead (not everyone has 32+ GB of RAM).

I thing it would be worth timing some more stuff here:
a) How much is the startup cost (e.g. before even startin the first resident compiler)?
b) How much time is actually spend compiling via the resident compilers? How much of that time is serializing?
c) How much time is actually spend starting up VMs? Waiting to be actually scheduled? Actually running the tests? (I don't know how much of this is actually possible to time)

Another thing to look into here would be if it would be possible to turn all of this into more of a hot reload like scenario: My guess would be that combined, much of the time is spend serializing, deserializing (and creating the classes etc in the VM) - the same thing again and again (say, the flutter framework itself). Would it be possible to just have the same VM load the flutter framework once, and then "reload" the "main script" (the actual test) and rerun?

(Also - in general - I think the test package should probably be updated to do concurrency differently. This of cause might actually already have been done; it's been a while since I looked at it. Setting concurrency to 1 but then actually having 2 VMs load and start compiling everything doesn't seem right I think.)

Anyways, I digress... I'm not sure what to think about this PR...

**Comment by aam** [comment]
Awesome input, thanks @jensjoha . Yeah, I need to get more data on this. It might be that amount of RAM is as critical(or even more so) than number of cores. I also just noticed that packages/flutter `flutter test` executed on CI didn't improve neither.
I would hate to leave things unchanged though because we seem to be leaving so much performance on the table for (above-average) case of high-powered desktop.

> Would it be possible to just have the same VM load the flutter framework once, and then "reload" the "main script" (the actual test) and rerun?

It would require some effort because VM in question is wrapped in flutter engine and running a test starts with initializing the engine and the VM, and then loading flutter app from compiled kernel file into that engine. Yes, engine is set up to do hot-reload/hot-restart for a given app, but I'm not sure whether we can run independent tests by swapping just dart part via hot-reload as I would imagine for flutter test to start properly engine might also have to be reinitialized.

**Comment by Hixie** [comment]
Should we close this PR while we look for additional information? (I'm trying to clean out the review queue.)

**Comment by aam** [comment]
Closing for investigation

**Pull request closed without merging**